finetune/base_models/bert/featurizer.py

```diff
@@ -148,3 +165,135 @@ def get_zeros():
+
+
+def table_roberta_featurizer_twinbert(
+    X,
+    encoder,
+    config,
+    train=False,
+    reuse=None,
+    context=None,
+    underlying_model=TwinBertModel,
+    max_length=None,
+    lengths=None,
+    **kwargs
+):
+    """
+    The backbone of the table model.
+
+    :param X: A tensor of token indexes with shape [batch_size, sequence_length, token_idx]
+    :param encoder: A TextEncoder object.
+    :param config: A config object, containing all parameters for the featurizer.
+    :param train: If this flag is true, dropout and losses are added to the graph.
+    :param reuse: Should reuse be set within this scope.
+    :return: A dict containing;
+        embed_weights: the word embedding matrix.
+        features: The output of the featurizer_final state.
+        sequence_features: The output of the featurizer at each timestep.
+    """
+    is_roberta = config.base_model.is_roberta
+    is_roberta_v1 = is_roberta and config.base_model.encoder == RoBERTaEncoder
+
+    if max_length is None:
+        max_length = config.max_length
+    bert_config = BertConfig(
+        vocab_size=encoder.vocab_size,
+        hidden_size=config.n_embed,
+        num_hidden_layers=config.n_layer,
+        num_attention_heads=config.n_heads,
+        intermediate_size=config.bert_intermediate_size,
+        hidden_act=config.act_fn,
+        hidden_dropout_prob=config.resid_p_drop,
+        attention_probs_dropout_prob=config.attn_p_drop,
+        max_position_embeddings=512,  # Overriden because the max length of the transformers is not the max length from our config.
+        type_vocab_size=2,
+        initializer_range=config.weight_stddev,
+        low_memory_mode=config.low_memory_mode,
+        pos_injection=config.context_injection,
+        reading_order_removed=config.reading_order_removed,
+        anneal_reading_order=config.anneal_reading_order,
+        positional_channels=config.context_channels,
+        table_position=config.table_position,
+        table_position_type=config.table_position_type,
+    )
+
+    initial_shape = tf.shape(input=X)
+    X = tf.reshape(X, shape=tf.concat(([-1], initial_shape[-1:]), 0))
+
+    if is_roberta:
+        # In our use case (padding token has index 1), roberta's position indexes begin at 2, so our
+        # positions embeddings come from indices 2:514.
+        if is_roberta_v1:
+            # v1 vocab didn't include MASK token although the embedding did
+            bert_config.vocab_size += 1
+
+        bert_config.max_position_embeddings += 2
+
+    if config.num_layers_trained not in [config.n_layer, 0]:
+        raise ValueError(
+            "Bert base model does not support num_layers_trained not equal to 0 or n_layer"
+        )
+
+    output_shape = tf.concat([tf.shape(X), [config.n_embed]], axis=0)
+
+    end_col, end_row, start_col, start_row = tf.unstack(context, num=4, axis=2)
+    row_gather = get_gather_indices(X, lengths, start_row, end_row)
```

Comment:
Record shape of output [num_rows, max_row_length, 2] where 2 references batch, seq index.